tests/run-make/remove-dir-all-race/rmake.rs

```diff
@@ -0,0 +1,37 @@
+use run_make_support::{
```

Comment:
This test slightly concerns me, as in is this a non-deterministic test that *usually* fails or *usually passes*, but we can guarantee it to be reproducible?

Reply:
This will also need to run through various try jobs, including various cross-compilation scenarios.

Is this test meant to be reproducible across each platform? Are macOS/various linux/Windows behavior different with respect to remove_dir_all?

Also note that some of the CI runners are ran as root, which may cause behavioral differences versus running as non-root. (As in, even if it passes in CI does not guarantee it passes locally as non-root...)

Reply:
I think to test this reliably it'd require a much larger directory structure and with many more threads. Even then it may not be 100% reliable.

Reply:
if we make the test platform specific we may be able to do some LD_PRELOAD fault injection? not sure if that's better^^

Reply:
I do not think that's better, lol.

Reply:
race conditions are inherently nondeterministic, so i don't think that can be helped.

it would be possible to insert yield_now calls specifically for testing, but then the code being tested isn't the same as that being used in production.

Reply:
I *think* @jieyouxu is suggesting that it may be better to have no test than a potentially flaky test? Because flaky tests can be hard to pin down and diagnose.

Reply:
> This will also need to run through various try jobs, including various cross-compilation scenarios.

can you point me to an existing test that does this so i can implement its strategies into my test?

> Is this test meant to be reproducible across each platform?

it should be, but that will probably need to be done by someone else, as i am not a windows programmer

> Are macOS/various linux/Windows behavior different with respect to remove_dir_all?

windows already ignores ENOENT when removing files (but not directories?), no idea how it works on macos

> I think to test this reliably it'd require a much larger directory structure and with many more threads. Even then it may not be 100% reliable.

a larger directory structure *might* help, but i'm not sure how i would integrate more threads into this.  because `remove_dir_all` still returns an error if the top-level dir is missing, only one of the threads can actually check for errors, while the other thread only deletes files within the outer dir, not the outer dir itself.

personally, i think the best way to decrease the chance of false negatives would be to run the existing test more times with smaller increments to the `thread::sleep` call.  the reason i didn't do this already is because i didn't want the test to take too long, and because the existing test already did a good job of catching errors (i implemented the test before the fix, so i know it fails without it)

Reply:
> it would be good to take a quick look at the other remove_dir_all implementations in library/std/src/sys/pal and see if there are any obvious similar situations.

thanks for the pointer, luckily these implementations were pretty simple so i could fix them without knowing a ton about the individual platforms.

Reply:
> I _think_ @jieyouxu is suggesting that it may be better to have no test than a potentially flaky test? Because flaky tests can be hard to pin down and diagnose.

To be clear, I'm not *advocating* to have no tests, I'm just uneasy whenever I see a *potentially* flakey test that may *potentially* non-deterministically fail (whether in CI or locally). I'm not going to block this PR on this (I'm not a T-libs reviewer anyway) especially since I don't know of a better alternative way to deterministic test this.

Reply:
I'm also somewhat uncomfortable with this test. I would prefer if the test didn't fail if it couldn't manage to cause a race.

Reply:
@Amanieu is there another status a run-make test can have other than pass or fail? i would like to alert people if the test breaks.

Reply:
Not that I know of. Either the test passes or it fails and blocks CI.

Reply:
> I'm also somewhat uncomfortable with this test. I would prefer if the test didn't fail if it couldn't manage to cause a race.

+1 to this. If this test absolutely *has* to be flaky and non-deterministic and has no other choice, then I would prefer it at least not fail if we couldn't manage to cause a race, because otherwise it can randomly block *every* PR on PR CI or full CI. It's still not ideal because then if this regresses, a PR-local try-job might pass, and so might its full CI test run, but then start failing in completely unrelated PRs.

> Is there another status a run-make test can have other than pass or fail? i would like to alert people if the test breaks.

This would require some significant infra support, maybe even moving to a new test suite called `run-make-no-block` that doesn't block full CI but is run in a separate CI job (something like a tool state check job?)... but then the feedback loop can be quite long.